### PR TITLE
Use a Garth OAuth implementation fixes #95

### LINKED ISSUE #95: Broke again? GarminException: Did not find the display name in the profile page.
Seems to be able to get the Ticket, but after that, it's not getting anything.

➜  garmin-connect-export-master python gcexport.py --count 1                                                                       
Welcome to Garmin Connect Exporter!
[WARNING] Output directory ./2023-09-26_garmin_connect_export already exists. Will skip already-downloaded files and append to the CSV file.
Username: username@gmail.com
Password: 
Connecting to Garmin Connect... Done.
Requesting Login ticket... Done. Ticket=ST-0426532-gzLYoa1LJbUhxVun3keA-cas
Authenticating... Done.
Getting display name...Traceback (most recent call last):
  File "/Users/username/Downloads/garmin-connect-export-master/gcexport.py", line 1349, in <module>
    main(sys.argv)
  File "/Users/username/Downloads/garmin-connect-export-master/gcexport.py", line 1301, in main
    userstats_json = fetch_userstats(args)
                     ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/username/Downloads/garmin-connect-export-master/gcexport.py", line 952, in fetch_userstats
    display_name = extract_display_name(profile_page)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/username/Downloads/garmin-connect-export-master/gcexport.py", line 977, in extract_display_name
    raise GarminException('Did not find the display name in the profile page.')
GarminException: Did not find the display name in the profile page.

### PR DESCRIPTION
**Pull request by SimonBaars**
For full context, see the relevant issue thread: https://github.com/pe-st/garmin-connect-export/issues/95

TLDR: Garmin shut down the old authentication flow. The new authentication flow has many security measures in place to script the login. Using [a recommendation in an issue on a similar repo](https://github.com/petergardfjall/garminexport/issues/102), we use the Garth library to implement an OAuth flow, which should be more robust than the original method.

Review notes:
- The OAuth flow also required using the new endpoints (which could be migrated by simply removing the proxy)
- Removed the user profile fetch, since it was really only to determine the number of activities, and there's a way simpler way to do that.

Thanks to @geraudloup for investigating this with me, and @app4g for the support!

### CODE REVIEW

**Comment by sgowtham** [comment]
I can confirm that the python script is working as expected after running the following commands

```
pip3 install garth --user
curl https://patch-diff.githubusercontent.com/raw/pe-st/garmin-connect-export/pull/96.diff | git apply
```

When I run

`python3 ./gcexport.py -c 1 -ot -u`


it works but the workflow continues to download far more than just one activity. I am assuming the final observation is a case of RTFM on my part.

**Comment by SimonBaars** [comment]
@sgowtham Thanks for testing, and that's not a RTFM on your part! I was a little bit overzealous when I re-designed the activity fetching 😧  
I committed a fix!

**Comment by sgowtham** [comment]
Thank you :)

This could be a separate feature request, if need be. I am wondering if all files associated with an activity (e.g., csv, fit, gpx, kml, tcx, and hrZones) could  be downloaded in one call instead of making separate calls for each such file?

**Comment by SimonBaars** [comment]
@sgowtham Whether it's technically possible, I'm not sure. But yeah, that should be a separate feature. Feel free to open an issue to start a discussion :)

**Comment by pe-st** [comment]
Thanks a lot for investigating this and coming up with a solution! This week I have no time to look at the PR in detail, but I hope I'll manage to early next eek

File: gcexport.py [hunk 1]
```diff
@@ -88,63 +89,28 @@
 # Maximum number of activities you can request at once.
 # Used to be 100 and enforced by Garmin for older endpoints; for the current endpoint 'URL_GC_LIST'
 # the limit is not known (I have less than 1000 activities and could get them all in one go)
-LIMIT_MAXIMUM = 1000
+SINGLE_REQUEST_LIMIT = 1000
+# The absolute limit for activities is 10.000 on server side
+TOTAL_ACTIVITY_LIMIT = 10_000
```
Comment (pe-st) [inline gcexport.py:94]: Can I suggest to revert the changes related to maximum number of activities? I tried using `--count all` in your branch and it tried to download only about 1000 activities, even though I have now more than 2000 activities (I know I should fix the comment on line 91 ;-))

File: gcexport.py [hunk 2]
```diff
@@ -963,50 +891,28 @@ def fetch_userstats(args):
...
-    display_name = match.group(1)
-    return display_name
-
-
-def fetch_activity_list(args, total_to_download):
+def fetch_activity_list(args):
     """
     Fetch the first 'total_to_download' activity summaries; as a side effect save them in json format.
     :param args:              command-line arguments (for args.directory etc)
-    :param total_to_download: number of activities to download
```
Comment (pe-st) [inline gcexport.py:986]: Same here about reverting the changes related to maximum number of activities.

File: gcexport.py [hunk 3]
```diff
@@ -963,50 +891,28 @@ def fetch_userstats(args):
...
+        num_to_download = min(SINGLE_REQUEST_LIMIT, total_to_download - num_downloaded)
+        chunk = fetch_activity_chunk(args, num_to_download, num_downloaded)
         activities.extend(chunk)
-        total_downloaded += num_to_download
+        if len(chunk) != SINGLE_REQUEST_LIMIT:
+            break
 
-    # it seems that parent multisport activities are not counted in userstats
-    if len(activities) != total_to_download:
-        logging.info('Expected %s activities, got %s.', total_to_download, len(activities))
```
Comment (pe-st) [inline gcexport.py:1009]: Same here about reverting the changes related to maximum number of activities.

File: gcexport.py [hunk 4]
```diff
@@ -1296,18 +1202,10 @@ def main(argv):
...
-    userstats_json = fetch_userstats(args)
-
-    if args.count == 'all':
-        total_to_download = int(userstats_json['userMetrics'][0]['totalActivities'])
-    else:
-        total_to_download = int(args.count)
-
-    device_dict = {}
+    # Get user stats
+    fetch_userstats(args)
```
Comment (pe-st) [inline gcexport.py:1206]: Same here about reverting the changes related to maximum number of activities.

File: gcexport.py [hunk 5]
```diff
@@ -1317,12 +1215,13 @@ def main(argv):
         write_to_file(os.path.join(args.directory, 'event_types.properties'), event_type_props, 'w')
     event_type_name = load_properties(event_type_props)
 
-    activities = fetch_activity_list(args, total_to_download)
+    activities = fetch_activity_list(args)
```
Comment (pe-st) [inline gcexport.py:1218]: Same here about reverting the changes related to maximum number of activities.

File: requirements.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+garth==0.4.28
```
Comment (pe-st) [inline requirements.txt:1]: It seems that 0.4.28 doesn't exist everywhere (I could install it locally, but [github action](https://github.com/pe-st/garmin-connect-export/actions/runs/6337921408/job/17322992456?pr=96) fails with `Could not find a version that satisfies the requirement garth==0.4.28 (from versions: 0.4.29, 0.4.30, 0.4.31, 0.4.32, 0.4.33, 0.4.34, 0.4.35, 0.4.36)`)

Is there a way in `requirements.txt` to specify "latest" or 0.4.x, as garth has seen a couple of releases since opening the PR ;-)

File: gcexport.py [hunk 7]
```diff
@@ -963,50 +891,28 @@ def fetch_userstats(args):
     return json.loads(result)
 
 
-def extract_display_name(profile_page):
```
Comment (pe-st) [inline gcexport.py:966]: I guess you'd need to remove also `test_extract_display_name` in gcexport_test.py

File: README.md [hunk 8]
```diff
@@ -44,6 +44,7 @@ If you have many activities, you may find that this script crashes with an "Oper
 - If you're comfortable using Git, just clone the repo from github
 - Otherwise get the latest `zip` (or `tar.gz`) from the [releases page](https://github.com/pe-st/garmin-connect-export/releases)
   and unpack it where it suits you.
+- Install the dependencies: `pip install -r requirements.txt`
```
Comment (pe-st) [inline README.md:47]: I had success with `python3 -m pip install -r requirements.txt`, but not with `pip install -r requirements.txt`

My python installation only has `/usr/bin/pip3`, no `pip`. I guess using `python3 -m pip` works for more people, what do you think?

**Review by pe-st — COMMENTED** [review]
Thanks for your great work :-) I have only some minor suggestions

File: README.md (hunk 8 above)
Comment (SimonBaars) [reply README.md:47]: Yeah, indeed, `pip install -r requirements.txt` depends on where your `pip` executable points to, `python3 -m ` uses the `python3` pip version, which is more likely to work. Will update.

**Comment by SimonBaars** [comment]
Hey, I addressed all comments!

**Comment by SimonBaars** [comment]
Hey @pe-st , I addressed all comments :)

**Comment by telemaxx** [comment]
i just tried this PR

when i enter:

`c:\Python3\python.exe C:\Users\top\git\garmin-connect-export-pest\gcexport.py -c 5 -f gpx --username "myname" --password "mypassword" --directory "C:\public\gps"`

all activities are downloaded. (Should be only 5)

do i something wrong?

**Comment by pe-st** [comment]
> i just tried this PR
> 
> when i enter:
> 
> `c:\Python3\python.exe C:\Users\top\git\garmin-connect-export-pest\gcexport.py -c 5 -f gpx --username "myname" --password "mypassword" --directory "C:\public\gps"`
> 
> all activities are downloaded. (because of `-c 5` it should be only 5)
> 
> @SimonBaars: do i something wrong?

I think this was a small glitch that I fixed when merging the PR

**Comment by mainzelM** [comment]
@SimonBaars: Thanks a lot for the fix!
Also thanks to  @pe-st for maintaining this repo :smile:

**Comment by telemaxx** [comment]
@SimonBaars @martin @pe-st : from my side also a big Thank You.

everything is working with windows and linux.

BTW: updating of garth with: `pip install --upgrade garth`
(maybe helpful for someone)
